### PR TITLE
Unique pallet

### PR DESCRIPTION
**Pull request by rohibhan**
Need a Unique pallet with the following functionalities.
- [x] Digest of the document/data
- [x] Authorisation ID support
- [x] Optional attribute - stream identifier
- [x] If this option is used, check the status and ownership of the identifier.
- [x] Add stream identifier link to the unique storage
- [x] Create a new identifier - use a new prefix.
- [x] Revoke / Restore functionality.

### CODE REVIEW

**Comment by amarts** [comment]
Can the build errors be fixed in this?

**Comment by rohibhan** [comment]
fixed errors and added revoke function

File: pallets/unique/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+[dependencies]
+codec = { package = "parity-scale-codec", version = "3.4.0", default-features = false, features = [
+	"derive",
+	"max-encoded-len",
+] }
+scale-info = { version = "2.5.0", default-features = false, features = ["derive"] }
+cord-primitives = { package = "cord-primitives", path="../../primitives", default-features = false }
+cord-utilities = { package = "cord-utilities", path = "../../utilities", default-features = false }
+pallet-registry = {  default-features = false, path = '../registry' }
+pallet-schema = {  default-features = false, path = '../schema' }
```
Comment (amarts) [inline pallets/unique/Cargo.toml]: This dependency is not required I guess. Similarly review every presence of 'Schema' in the PR.

File: pallets/unique/src/types.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+/// An on-chain stream entry mapped to an Identifier.
+/// `UniqueEntry` is a struct that contains a `StreamDigestOf`, a `StreamCreatorIdOf`, a `SchemaIdOf`, a
+/// `RegistryIdOf`, and a `StatusOf`.
+///
+/// Properties:
+///
+/// * `digest`: The hash of the stream.
+/// * `creator`: The account that created the stream.
+/// * `schema`: The schema identifier.
```
Comment (amarts) [inline pallets/unique/src/types.rs]: wrong doc, remove schema

File: primitives/src/curi.rs [hunk 3]
```diff
@@ -75,6 +77,8 @@ pub enum IdentifierType {
 	Authorization,
 	Schema,
 	Stream,
+	//TODO : added unique in identifire type
```
Comment (amarts) [inline primitives/src/curi.rs]: is it still a TODO?

also minor nitpick: 's/identifire/identifier/g'

**Comment by amarts** [comment]
Plan to use it in one of the demo script by end of the month. Please focus on getting the tests passed, and lets get this merged.

**Comment by rohibhan** [comment]
> Plan to use it in one of the demo script by end of the month. Please focus on getting the tests passed, and lets get this merged.

Sure plannig to complete it by end of this month
Thanks

**Comment by amarts** [comment]
Can you rebase it on top of develop ?

File: pallets/unique/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	/// unique Transaction stored on chain.
+	/// It maps from a unique transaction to an identifier (resolve from hash).
+	#[pallet::storage]
+	#[pallet::getter(fn unique_identifiers)]
+	pub type UniqueIdentifiers<T> =
+		StorageMap<_, Blake2_128Concat, InputUniqueOf<T>, UniqueIdOf, OptionQuery>;
```
Comment (smohan-dw) [inline pallets/unique/src/lib.rs:124]: Storage here should be named as UniqueDigestEntries, as it's mapping the input to the identifier.

File: pallets/unique/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+
+	/// unique hashes stored on chain.
+	/// It maps from a unique hash to an metadata (resolve from hash).
+	#[pallet::storage]
+	#[pallet::getter(fn unique_digest_entries)]
+	pub type UniqueDigestEntries<T> =
```
Comment (smohan-dw) [inline pallets/unique/src/lib.rs]: This one should be UniqueIdentifiers as it is mapping the Identifier to details.

File: pallets/unique/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A new unique identifier has been created.
+		/// \[unique identifier, unique digest, controller\]
+		Create { identifier: UniqueIdOf, digest: InputUniqueOf<T>, author: UniqueCreatorIdOf<T> },
+		/// A unique identifier status has been revoked.
+		/// \[unique identifier, controller\]
+		Revoke { identifier: UniqueIdOf, author: UniqueCreatorIdOf<T> },
```
Comment (smohan-dw) [inline pallets/unique/src/lib.rs:157]: missing update

File: pallets/unique/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+				UniqueCommitActionOf::Genesis,
+			)
+			.map_err(<Error<T>>::from)?;
+
+			// Emit a Create event
+			Self::deposit_event(Event::Create { identifier, digest: unique_txn, author: creator });
+
+			Ok(())
+		}
+		/// Revokes a unique.
```
Comment (smohan-dw) [inline pallets/unique/src/lib.rs:385]: With an identifier-led approach, having support for updating the identifier with an updated digest is good.

File: pallets/unique/src/types.rs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	pub committed_by: UniqueCreatorIdOf,
+	/// Stream block number
+	pub created_at: Timepoint<BlockNumber>,
+}
+
+//Question to adsk use of identifier ?
+
+/// Defining the possible actions that can be taken on a stream.
+#[derive(Clone, Copy, RuntimeDebug, Decode, Encode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub enum UniqueCommitActionOf {
```
Comment (smohan-dw) [inline pallets/unique/src/types.rs:84]: Remove the digest from this. need to have a discussion on other values that are not supported atm

**Comment by rohibhan** [comment]
> Can you rebase it on top of develop ?

**Comment by rohibhan** [comment]
closing PR as new branch named unique_pallet_sc is created

**Pull request closed without merging**